lib/p2p/Peer.ts

```diff
@@ -255,6 +422,102 @@ class Peer extends EventEmitter {
 
     this.emit('error', { msg, err });
   }
+
+  private sendHello = (): HelloPacket => {
+    // TODO: use real values
+    const packet = new HelloPacket({
+      version: '123',
+      nodeKey: '123',
+      listenPort: 20000,
+      pairs: ['BTC/LTC'],
+    });
+
+    this.sendPacket(packet);
+
+    return packet;
+  }
+
+  private handleHello = (packet: HelloPacket): void => {
+    this.handshakeState = { ...this.handshakeState, ...packet.body };
+
+    // TODO: define XUD_MIN_VERSION, throw exception if version below
+  }
+
+  private sendPing = (): PingPacket => {
+    this.lastPingId = uuidv1();
+    const packet = new PingPacket({ id: this.lastPingId! });
+
+    this.sendPacket(packet);
+
+    return packet;
+  }
+
+  private handlePing = (packet: PingPacket): void  => {
+    const { id } = packet.body;
+
+    if (!id) {
+      return;
+    }
+
+    this.sendPong(id + 1);
+  }
+
+  private sendPong = (id: string): PongPacket => {
+    const packet = new PongPacket({ id });
+
+    this.sendPacket(packet);
+
+    return packet;
+  }
+
+  private handlePong = (packet: PongPacket): boolean => {
+    const { id } = packet.body;
+    if (!this.lastPingId) {
+      this.logger.debug(`Peer sent an unsolicited pong (${this.id})`);
+      return false;
+    }
+
+    if (id !== this.lastPingId) {
+      this.logger.debug(`Peer sent an invalid pong id (${this.id})`);
+      return false;
+    }
+
+    this.lastPingId = null;
+    return true;
+  }
+}
+
+class ResponseEntry {
+  public timeout: number = 0;
+  public jobs: Job[] = [];
+
+  public addJob = (resolve, reject) => {
+    this.jobs.push(new Job(resolve, reject));
+  }
+
+  public setTimeout = (timeout: number): void => {
+    this.timeout = ms() + timeout;
+  }
+
+  public resolve = (result: any) => {
+    for (const job of this.jobs) {
+      job.resolve(result);
+    }
+
+    this.jobs.length = 0;
+  }
+
+  public reject = (err: any) => {
+    for (const job of this.jobs) {
+      job.reject(err);
+    }
+
+    this.jobs.length = 0;
+  }
+}
+
+class Job {
```

Comment:
How will this be used differently from a regular promise?

Reply:
It's actually equivalent to a promise executor, not to a promise (which you can call `then` on). It's being used just to resolve promises, not to define/hold promises. We might need to add that later, but I didn't find a use-case for that yet.

Reply:
Ok, what do you think about calling this `Resolver` or something along those lines?